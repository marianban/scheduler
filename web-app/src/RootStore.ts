import { AppointmentsModel } from 'appointments/AppointmentsModel';
import { CalendarStore } from 'calendar/CalendarStore';
import { DateSelectionModel } from 'calendar/DateSelectionModel';
import { ClientSelectionModel } from 'clients/ClientSelectionModel';
import { ClientStore } from 'clients/ClientStore';
import { PubSub } from 'PubSub';

export class RootStore {
  public calendarStore: CalendarStore;
  public clientStore: ClientStore;
  public appointmentsModel: AppointmentsModel;
  public dateSelectionModel: DateSelectionModel = new DateSelectionModel();
  public pubSub: PubSub = new PubSub();
  public clientSelectionModel: ClientSelectionModel;

  constructor(currentDate: Date) {
    this.clientStore = new ClientStore();
    this.appointmentsModel = new AppointmentsModel();
    this.clientSelectionModel = new ClientSelectionModel();
    this.calendarStore = new CalendarStore(currentDate);
    this.clientStore.onClientDeleted(this.cancelDeletedClientAppointments);
    this.clientStore.onClientDeleted(this.unselectDeletedClient);
  }

  private cancelDeletedClientAppointments = (clientId: string) => {
    this.appointmentsModel.cancelByClientId(clientId);
  };

  private unselectDeletedClient = (clientId: string) => {
    if (
      this.clientSelectionModel.selectedClient !== null &&
      this.clientSelectionModel.selectedClient.id === clientId
    ) {
      if (this.clientStore.clients.length > 0) {
        this.clientSelectionModel.select(this.clientStore.clients[0]);
      } else {
        this.clientSelectionModel.unselect();
      }
    }
  };
}
