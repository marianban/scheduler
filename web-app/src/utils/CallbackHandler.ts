export type UnsubscribeCallback = () => void;

export class CallbackHandler<T> {
  private callbacks: Array<(item: Readonly<T>) => void> = [];

  public add(callback: (item: Readonly<T>) => void): UnsubscribeCallback {
    this.callbacks.push(callback);
    const unsubscribe = () => {
      const index = this.callbacks.indexOf(callback);
      if (index !== -1) {
        this.callbacks.splice(index, 1);
      }
    };
    return unsubscribe;
  }

  public handle(item: T) {
    this.callbacks.forEach(callback => {
      callback(item);
    });
  }
}
