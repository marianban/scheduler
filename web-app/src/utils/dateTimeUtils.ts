import format from "date-fns/format";
import parse from "date-fns/parse";

type Precision =
  | "milliseconds"
  | "seconds"
  | "minutes"
  | "hours"
  | "days"
  | "days"
  | "month"
  | "year";

const getArg = (args: number[], index: number) => args[index] || 0;

export const toPrecision = (date: Date, precision: Precision) => {
  const precisionMap = {
    milliseconds: 7,
    seconds: 6,
    minutes: 5,
    hours: 4,
    days: 3,
    month: 2,
    year: 1
  };
  const parts = [
    date.getFullYear(),
    date.getMonth(),
    date.getDate(),
    date.getHours(),
    date.getMinutes(),
    date.getSeconds(),
    date.getMilliseconds()
  ];
  const end = precisionMap[precision];
  const args = parts.slice(0, end);
  return new Date(
    getArg(args, 0),
    getArg(args, 1),
    getArg(args, 2),
    getArg(args, 3),
    getArg(args, 4),
    getArg(args, 5),
    getArg(args, 6)
  );
};

export const parseDate = (value: string, baseDate: Date = new Date()) => {
  const matches = value.match(/(\d+)[\.\-\/]?(\d*)[\.\-\/]?(\d*)/);
  if (!matches) {
    throw new Error(`Unable to parse date: ${value}`);
  }
  const [day, month, year] = matches.slice(1);
  const date = `${day}${month && "."}${month}${year && "."}${year}`;
  const pattern = `dd${month && ".MM"}${year && ".yyyy"}`;
  return parse(date, pattern, toPrecision(baseDate, "days"));
};

export const normalizeDate = (value: string, baseDate: Date) => {
  return format(parseDate(value, baseDate), "d/M/yyyy");
};

export const normalizeTime = (value: string) => {
  const matches = value.match(/(\d+):?(\d+)?/);
  if (!matches) {
    throw new Error(`Unable to parse time: ${value}`);
  }
  const [hour, minute] = matches.slice(1);
  return `${hour}:${minute || "00"}`;
};
