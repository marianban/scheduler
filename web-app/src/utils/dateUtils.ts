import { parse } from 'date-fns';

type Precision =
  | 'milliseconds'
  | 'seconds'
  | 'minutes'
  | 'hours'
  | 'days'
  | 'days'
  | 'month'
  | 'year';

export const toPrecision = (date: Date, precision: Precision) => {
  console.log('to precision');
  const precisionMap = {
    milliseconds: 7,
    seconds: 6,
    minutes: 5,
    hours: 4,
    days: 3,
    month: 2,
    year: 1
  };
  const parts = [
    date.getFullYear(),
    date.getMonth(),
    date.getDate(),
    date.getHours(),
    date.getMinutes(),
    date.getSeconds(),
    date.getMilliseconds()
  ];
  const end = precisionMap[precision];
  return new Function.prototype.bind.apply(
    Date,
    ...[null].concat(parts.slice(0, end))
  );
};

export const parseDate = (value: string, baseDate: Date = new Date()) => {
  const matches = value.match(/(\d+)[\.\-\/](\d+)[\.\-\/](\d+)/);
  if (!matches) {
    throw new Error(`Unable to parse: ${value}`);
  }
  const [day, month, year] = matches.slice(1);
  return parse(value, `${day}.${month}.${year}`, toPrecision(baseDate, 'days'));
};
