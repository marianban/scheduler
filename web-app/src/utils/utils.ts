export const someOrEmptyString = (value: string | null | undefined) =>
  value ? value : '';

export const callCallbacks = <T>(
  callbacks: Array<(item: Readonly<T>) => void>,
  item: T
) => {
  callbacks.forEach(callback => {
    callback(item);
  });
};
