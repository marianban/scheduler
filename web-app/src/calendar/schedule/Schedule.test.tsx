import { getByTestId } from 'dom-testing-library';
import * as React from 'react';
import { RootStore } from 'RootStore';
import { renderWithProviders } from 'test/utils';
import {
  getAppointmentPosition,
  getStartOfWorkDay,
  Schedule
} from './Schedule';

const renderSchedule = (rootStore: RootStore) => {
  return {
    ...renderWithProviders(<Schedule rootStore={rootStore} />, rootStore),
    rootStore
  };
};

it('renders a week for selected date', () => {
  const rootStore = new RootStore(new Date());
  rootStore.dateSelectionModel.set(new Date(2018, 10, 22));
  const { getAllByTestId }: any = renderSchedule(rootStore);
  const dayNumbers = getAllByTestId('week-day-name')
    .slice(1)
    .map((e: HTMLElement) => ({
      name: getByTestId(e, 'day-name').textContent,
      number: getByTestId(e, 'day-number').textContent
    }));
  expect(dayNumbers).toHaveLength(7);
  expect(dayNumbers).toMatchSnapshot();
});

it('renders new appointment on an appropriate place', () => {
  const rootStore = new RootStore(new Date());
  rootStore.dateSelectionModel.set(new Date(2018, 10, 22));
  rootStore.appointmentsModel.create({
    date: '20/11/2018',
    time: '10:00'
  });
  const { getAllByTestId }: any = renderSchedule(rootStore);
  const appointments = getAllByTestId('appointment');
  expect(appointments).toHaveLength(1);
});

it.each`
  date                     | row  | column
  ${'2018-10-29T08:00:00'} | ${2} | ${2}
  ${'2018-10-29T08:30:00'} | ${3} | ${2}
  ${'2018-10-29T08:31:00'} | ${3} | ${2}
  ${'2018-10-29T08:59:00'} | ${3} | ${2}
  ${'2018-10-29T09:00:00'} | ${4} | ${2}
  ${'2018-10-30T08:00:00'} | ${2} | ${3}
  ${'2018-11-01T08:00:00'} | ${2} | ${5}
`('for $date returns row: $row, column: $column', ({ date, row, column }) => {
  const jsDate = new Date(date);
  const result = getAppointmentPosition(getStartOfWorkDay(jsDate), jsDate);
  expect(result.row).toBe(row);
  expect(result.column).toBe(column);
});
