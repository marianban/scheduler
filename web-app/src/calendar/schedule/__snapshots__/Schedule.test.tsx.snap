// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a week for selected date 1`] = `
Array [
  Object {
    "name": "Mon",
    "number": "19",
  },
  Object {
    "name": "Tue",
    "number": "20",
  },
  Object {
    "name": "Wed",
    "number": "21",
  },
  Object {
    "name": "Thu",
    "number": "22",
  },
  Object {
    "name": "Fri",
    "number": "23",
  },
  Object {
    "name": "Sat",
    "number": "24",
  },
  Object {
    "name": "Sun",
    "number": "25",
  },
]
`;
